### PR TITLE
[ISSUE #1156]new mqadmin API for ACL configuration

### PR DESCRIPTION
**Pull request by zongtanghu**
## What is the purpose of the change

(1)Adding data new version mechanism for every broker's acl config file in a cluster.
(2)Adding ClusterAclConfigVersionListSubCommand for querying acl config version;
(3)Adding DeleteAccessConfigSubCommand for deleting account attribute in acl config file.
(4)Adding UpdateAccessConfigSubCommand for updating account attribute in acl config file.
(5)Adding UpdateGlobalWhiteAddrSubCommand for updating global white addrs in acl config file.

## Brief changelog

(1)Adding data new version mechanism for every broker's acl config file in a cluster.
(2)Adding ClusterAclConfigVersionListSubCommand for querying acl config version;
(3)Adding DeleteAccessConfigSubCommand for deleting account attribute in acl config file.
(4)Adding UpdateAccessConfigSubCommand for updating account attribute in acl config file.
(5)Adding UpdateGlobalWhiteAddrSubCommand for updating global white addrs in acl config file.

Adding data new version mechanism in the acl yaml configuration file for acl mqadmin for acl configuration.

## Verifying this change

I have already verified this changes in my local enviroment and passed acl uniting test cases.

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by zongtanghu** [comment]
Hi @dongeforever @duhenglucky I have pushed initial version of acl configuration command's implementation.You can review the code firstly.And I will verify this feature in this week.If I find some small issues,I will push other commits codes to repo.

**Comment by zongtanghu** [comment]
I have alreay verified some case below.For example,
(1)Execute the updateAclConfig mqadmin command to update some acl config element:
sh mqadmin updateAclConfig -n XX.XX.XX.XX -c DefaultCluster -a RocketMQ -s 1234567809123 -t topicA=DENY,topicD=SUB -g groupD=DENY,groupB=SUB
(2)Excuete the clusterAclConfigVersion to query the acl config file's version:
sh mqadmin clusterAclConfigVersion -n XX.XX.XX.XX:9876 -b XX.XX.XX.XX:10911
(3)Execute the updateAclConfig mqadmin command to create some acl config element:
sh mqadmin updateAclConfig -n XX.XX.XX.XX:9876 -b XX.XX.XX.XX:10911 -a RocketMQ -s 123456780912345 -w 192.168.0.* -t topicA=DENYm,topicD=SUB

**Comment by zongtanghu** [comment]
acl_command_1

**Comment by zongtanghu** [comment]
acl_command_2_after_deleting_aclconfig

**Comment by zongtanghu** [comment]
acl_command_3_updating_aclconfig

**Comment by zongtanghu** [comment]
acl_command_4_getAclVersionCommand

**Comment by zongtanghu** [comment]
image

**Comment by zongtanghu** [comment]
I have already verified these acl mqadmin command in our testing enviroment. @duhenglucky @dongeforever Please help to review this PR.Thanks!

**Comment by zongtanghu** [comment]
acl_command_5_getAclVersionCommand_in_Cluster

**Comment by zongtanghu** [comment]
acl_command_6_updating_aclconfig_in_Cluster

**Comment by zongtanghu** [comment]
@githublaohu @walking98 Please help to review this pr.If there is no problem.please comments with "LGTM".Thanks.

**Comment by walking98** [comment]
1. I suppose it will add doc/README to introduce the usage of the new mqadmin API/cmd.
2. I can not find an API to get the whole ACL content for a given broker, it is important for user/console to get the overview of the ACL at the very beginning.

**Comment by zongtanghu** [comment]
> 1. I suppose it will add doc/README to introduce the usage of the new mqadmin API/cmd.
> 2. I can not find an API to get the whole ACL content for a given broker, it is important for user/console to get the overview of the ACL at the very beginning.

(1)I will add the contents into the doc/acl_guide files.
(2)I will add the query acl yaml config file mqadmin command later.
Thank you very much, @walking98 suggestion for this pr.

**Comment by githublaohu** [comment]
LGTM

File: acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java [hunk 1]
```diff
@@ -89,10 +91,182 @@ public void load() {
...
+        if (StringUtils.isEmpty(plainAccessConfig.getAccessKey()) ||
+            plainAccessConfig.getAccessKey().length() <= 6) {
+            throw new AclException(String.format(
+                    "The accessKey=%s cannot be null and length should longer than 6",
+                    plainAccessConfig.getAccessKey()));
+        }
+        newAccountsMap.put(AclConstants.CONFIG_ACCESS_KEY, plainAccessConfig.getAccessKey());
+
+        if (!StringUtils.isEmpty(plainAccessConfig.getSecretKey())) {
+            if (plainAccessConfig.getSecretKey().length() <= 6) {
```
Comment (duhenglucky) [inline acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java]: Try not to use magic numbers, yeah, I mean 6, maybe it just is a username in some situation.

File: acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java [hunk 2]
```diff
@@ -89,10 +91,182 @@ public void load() {
...
+        if (plainAccessConfig.getGroupPerms() != null && !plainAccessConfig.getGroupPerms().isEmpty()) {
+            newAccountsMap.put(AclConstants.CONFIG_GROUP_PERMS, plainAccessConfig.getGroupPerms());
+        }
+
+        return newAccountsMap;
+    }
+
+    public boolean deleteAccessConfig(String accesskey) {
+        if (StringUtils.isEmpty(accesskey)) {
+            log.error("parameter value accesskey is null or empty String,please check your parameter");
```
Comment (duhenglucky) [inline acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java]: Capitalization the first letter is a good choice, and other places are also,

**Comment by duhenglucky** [comment]
@zongtanghu and it seems that the PlainPermissionLoader class has been become an ACL manage service, so how about rename this class to AclManage or other names?

**Comment by zongtanghu** [comment]
> @zongtanghu Please make sure the ci test has been passed in your local environment.

Yes,I passed all the ci tests in my local enviroment.

File: acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java [hunk 3]
```diff
@@ -89,10 +91,182 @@ public void load() {
             }
         }
 
+        //for loading dataversion part just
+        JSONArray dataVersion4Yaml = plainAclConfData.getJSONArray(AclConstants.CONFIG_DATA_VERSION);
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java]: Okay,I will modify this variable name.

**Comment by zongtanghu** [comment]
> @zongtanghu and it seems that the PlainPermissionLoader class has been become an ACL manage service, so how about rename this class to AclManage or other names?

Yes,I thinks so.I will rename this class in this pr.

File: acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java (hunk 2 above)
Comment (zongtanghu) [reply acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java]: Okay,I will modify this part.

File: acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java (hunk 1 above)
Comment (zongtanghu) [reply acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java]: Okay,I will put this constant value into a constant class field.

File: acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java [hunk 4]
```diff
@@ -89,10 +91,182 @@ public void load() {
...
+        Map<String, Object> aclAccessConfigMap = AclUtils.getYamlDataObject(fileHome + File.separator + fileName,
+            Map.class);
+
+        List<Map<String, Object>> accounts = (List<Map<String, Object>>) aclAccessConfigMap.get(AclConstants.CONFIG_ACCOUNTS);
+        Map<String, Object> updateAccountMap = null;
+        if (accounts != null) {
+            for (Map<String, Object> account : accounts) {
+                if (account.get(AclConstants.CONFIG_ACCESS_KEY).equals(plainAccessConfig.getAccessKey())) {
+                    //update acl access config elements
+                    accounts.remove(account);
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java]: Okay,I will optimize this part of codes.

**Comment by glinggit** [comment]
LGTM

**Comment by duhenglucky** [comment]
LGTM

File: acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (githublaohu) [reply acl/src/main/java/org/apache/rocketmq/acl/plain/PlainPermissionLoader.java]: You don't have to delete it, AclUtils.writeDataObject2Yaml and then load it

**Comment by vongosling** [comment]
@zongtanghu Adding data new version mechanism in the acl yaml configuration file for acl mqadmin for acl configuration.

I paste the description into the google translate, It's unlucky, I could not figure out what's the meaning

File: acl/src/main/java/org/apache/rocketmq/acl/AccessValidator.java [hunk 5]
```diff
@@ -35,4 +37,32 @@
      * @param accessResource
      */
     void validate(AccessResource accessResource);
+
+    /**
+     * update the access resource config
```
Comment (vongosling) [inline acl/src/main/java/org/apache/rocketmq/acl/AccessValidator.java]: The first letter needs to be capitalized. The first letter needs to be capitalized. The first letter needs to be capitalized. Three times make help to remember.

**Comment by vongosling** [comment]
@zongtanghu Another thing, If I could figure out what's you have doing in this pr, I would like to help to merge the pr. But. please remember, consider this change almost over 40 files. It's not a good practice to review and merge model. could you make a  smaller pr, just less than 10 files affected in the next time?

File: docs/cn/acl/user_guide.md [hunk 6]
```diff
@@ -82,5 +82,76 @@ RocketMQ的权限控制存储的默认实现是基于yml配置文件。用户可
...
+说明：如果指定的是集群名称，则会在集群中各个broker节点执行该命令；否则会在单个broker节点执行该命令。
+
+| 参数 | 取值 | 含义 |
+| --- | --- | --- |
+| n | eg:192.168.1.2:9876 | namesrv地址 |
+| c | eg:DefaultCluster | 指定集群名称(与broker地址二选一) |
+| b | eg:192.168.12.134:10911 | 指定broker地址(与集群名称二选一) |
+
+
 **特别注意**在[4.5.0]版本中即使使用上面所述的白名单也无法解决开启ACL的问题，解决该问题的[PR链接](https://github.com/apache/rocketmq/pull/1149)
```
Comment (vongosling) [inline docs/cn/acl/user_guide.md]: 擦，可以直接说明，这个问题在4.5.1修复就好了，详细pr见这里。否则同学们不知道这个合并到主干没

**Comment by zongtanghu** [comment]
> @zongtanghu Another thing, If I could figure out what's you have doing in this pr, I would like to help to merge the pr. But. please remember, consider this change almost over 40 files. It's not a good practice to review and merge model. could you make a smaller pr, just less than 10 files affected in the next time?

Okay,This PR is too much large.I will remember the rule of puting forward pr for rocketmq source community.

File: docs/cn/acl/user_guide.md [hunk 7]
```diff
@@ -82,5 +82,76 @@ RocketMQ的权限控制存储的默认实现是基于yml配置文件。用户可
...
+| n | eg:192.168.1.2:9876 | namesrv地址 |
+| c | eg:DefaultCluster | 指定集群名称(与broker地址二选一) |
+| b | eg:192.168.12.134:10911 | 指定broker地址(与集群名称二选一) |
+| a | eg:RocketMQ | Access Key值 |
+| s | eg:1234567809123 | Secret Key值 |
+| m | eg:true | 是否管理员账户 |
+| w | eg:192.168.0.* | whiteRemoteAddress,用户IP白名单 |
+| i | eg:SUB | defaultTopicPerm,默认Topic权限 |
+| u | eg:PUB | defaultGroupPerm,默认ConsumerGroup权限 |
+| t | eg:topicA=DENY,topicD=SUB | topicPerms,各个Topic的权限 |
```
Comment (vongosling) [inline docs/cn/acl/user_guide.md]: 这里到底有几个值呢，deny，pub，sub？文档还是要仔细些

**Comment by zongtanghu** [comment]
> @zongtanghu Adding data new version mechanism in the acl yaml configuration file for acl mqadmin for acl configuration.
> 
> I paste the description into the google translate, It's unlucky, I could not figure out what's the meaning.
> . I think you could detail your pr purpose, more guys maybe help to review and it's a better way to be let more guys know about what you have done according to your simple text.

Adding data new version mechanism in the acl config file is helping to view different version of every broker's file for users.Because every broker persist its acl config file into local acl config file. @vongosling

**Comment by zongtanghu** [comment]
@vongosling I have already disscussed the data new version mechanism with @duhenglucky @dongeforever @walking98  in a issue,the link is: https://github.com/apache/rocketmq/issues/1156

File: docs/cn/acl/user_guide.md (hunk 7 above)
Comment (zongtanghu) [reply docs/cn/acl/user_guide.md]: Okay,I will write this contants much more carefully.

File: docs/cn/acl/user_guide.md (hunk 6 above)
Comment (zongtanghu) [reply docs/cn/acl/user_guide.md]: Okay,I describe this sentance again.

File: docs/cn/acl/user_guide.md (hunk 7 above)
Comment (vongosling) [reply docs/cn/acl/user_guide.md]: yeap, when we write the configuration, we must be careful what's is the default, what's is required, what's is optional value.

**Comment by zongtanghu** [comment]
Hi @vongosling  I have already resolve the issue that you mentioned above.

**Comment by vongosling** [comment]
@zongtanghu Could you help to rebase the pr? I found too many commits for different resolved issue.

File: docs/cn/acl/user_guide.md [hunk 8]
```diff
@@ -82,5 +82,76 @@ RocketMQ的权限控制存储的默认实现是基于yml配置文件。用户可
...
+## 7. ACL mqadmin配置管理命令
+
+### 7.1 更新ACL配置文件中“account”的属性值
+
+该命令的示例如下：
+
+sh mqadmin updateAclConfig -n 192.168.1.2:9876 -b 192.168.12.134:10911 -a RocketMQ -s 1234567809123 
+-t topicA=DENY,topicD=SUB -g groupD=DENY,groupB=SUB
+
+说明：如果不存在则会在ACL Config YAML配置文件中创建；若存在，则会更新对应的“accounts”的属性值;
```
Comment (jonnxu) [inline docs/cn/acl/user_guide.md]: 7.1标题中是account属性值，这里是accounts属性值？

File: docs/cn/acl/user_guide.md [hunk 9]
```diff
@@ -82,5 +82,76 @@ RocketMQ的权限控制存储的默认实现是基于yml配置文件。用户可
...
+如果指定的是集群名称，则会在集群中各个broker节点执行该命令；否则会在单个broker节点执行该命令。
+
+| 参数 | 取值 | 含义 |
+| --- | --- | --- |
+| n | eg:192.168.1.2:9876 | namesrv地址 |
+| c | eg:DefaultCluster | 指定集群名称(与broker地址二选一) |
+| b | eg:192.168.12.134:10911 | 指定broker地址(与集群名称二选一) |
```
Comment (jonnxu) [inline docs/cn/acl/user_guide.md]: namesrv和broker都举一个两个ip地址的例子，因为有些是用","隔开的，有些是用";"隔开的；broker地址是否可以同时指定多个？

File: docs/cn/acl/user_guide.md (hunk 8 above)
Comment (zongtanghu) [reply docs/cn/acl/user_guide.md]: Okay,I will adjust the content.

File: docs/cn/acl/user_guide.md (hunk 9 above)
Comment (zongtanghu) [reply docs/cn/acl/user_guide.md]: No,here,users can specify one broker to update the acl yaml config file rather than multiple ones.

**Pull request merged**